### PR TITLE
Init CI for 8.6.5, 8.8.4, 8.10.2

### PR DESCRIPTION
**Pull request by LisannaAtGalois**
~~All packages are built and all package tests except `crucible-llvm-tests` are run. I'll do a followup PR(s) for adding 8.10.2 and enabling `crucible-llvm-tests`, both of which currently have issues to resolve.~~

Build and test for `semmc-aarch32`, `semmc-arm`, `semmc-ppc`, and `semmc-x86_64`.

### CODE REVIEW

**Comment by LisannaAtGalois** [comment]
@kquick fortunately it looks like there aren't issues with building all the packages and running their tests. Just had to bump the what4 version to deal with the version bound issue that also came up in saw-script's travis CI.

**Comment by kquick** [comment]
Nice progress so far, thanks! A couple of things I'd like to explore:

* Can we build modules independently?  Possibly separate github actions?  This repository contains several related modules, but they could be considered individually distinct.
  1. If I'm working on semmc but not ready to make corresponding changes in the downstream (e.g semmc-ppc), the build will fail in semmc-ppc even though my changes in semmc might be building successfully.  Arguably the combined check would help in keeping the modules synchronized, but realistically I think there are projects that may only fund/need development for a specific architecture and so the synchronization of the other architectures might need to be deferred.
  2. If I'm working on a downstream module (e.g. semmc-ppc), a change there will cause the github actions to rebuild all modules, which adds considerably to the amount of time needed to run even though I haven't been changing those other modules.  I realize that github action caching will mitigate this to some degree, but I also think there are some fairly tight limits on the amount of cache  we can expect to be helping in this regard.

* Can you add in the other modules in this repo (preferrably separately per the above): semmc-coverage, semmc-fuzzer, semmc-toy, semmc-synthesis, and semmc-learning.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+          key: cabal-${{ matrix.package }}-${{ matrix.ghc }}-${{ hashFiles('**/cabal.GHC-*') }}-${{ github.sha }}
+          restore-keys: |
+            cabal-${{ matrix.package }}-${{ matrix.ghc }}-${{ hashFiles('**/cabal.GHC-*') }}
+
+      - name: Configure
+        run: |
+          ln -sv cabal.project.newbuild cabal.project
+          ln -sv cabal.GHC-${{ matrix.ghc }}.config cabal.project.freeze
+          cabal v2-update
+          cabal v2-configure -j2 --minimize-conflict-set
```
Comment (kquick) [inline .github/workflows/ci.yml:54]: Probably want to add `--enable-tests` to this line to ensure tests are built.

Comment (LisannaAtGalois) [reply .github/workflows/ci.yml]: Looks like tests for semmc have some issues :(. I've selectively disabled them in this PR so they can be handled separately.

Comment (LisannaAtGalois) [reply .github/workflows/ci.yml]: https://github.com/GaloisInc/semmc/pull/57 was opened for handling the failing semmc tests. This PR should be good to go in if you're happy with what's enabled here so far.

**Comment by kquick** [comment]
Thanks for separating out the different module builds.  The only downside I see now is that it's still going to try to build them all, even if only one of the modules has changes.  Is there any way with github actions to only run the build for the portion of the repository that has changed (and the downstream of that)?

**Comment by LisannaAtGalois** [comment]
> Thanks for separating out the different module builds. The only downside I see now is that it's still going to try to build them all, even if only one of the modules has changes. Is there any way with github actions to only run the build for the portion of the repository that has changed (and the downstream of that)?

@kquick Not easily. GitLab has built in support for that via the [`only:changes`](https://docs.gitlab.com/ee/ci/yaml/#onlychangesexceptchanges) rule, and it allows you to set a job's run condition to be based on files changed by the triggering **push event**. GitHub Actions doesn't have a corresponding rule for its job conditionals. I've seen people try to replicate it with git scripts, but these involve very fragile assumptions about what's being pushed (like that only one commit gets pushed at a time). GitLab is able to provide this safely because it keeps track of what changes are in each **push event**, which may contain multiple commits.

For the hypothetical case you've described where only a subset of the total work necessary to implement and integrate a change to `semmc` is funded, instead of trying to make the `master` branch CI tolerant of regressions to try to fit a theorized future need, I would suggest that such work be done on a branch, with the CI adjusted for that branch to capture the build needs defined by the project, thus allowing that branch to be the deliverable rather than a regressed `master`. If the project isn't defining the need to fully integrate the changes so that the repository builds cleanly, it doesn't make sense to me to also have those done on `master`.

In terms of turnaround time, splitting the build job out by a product of GHC versions and components provides a slight speed advantage for turnaround of specific components, but at a cost of much larger build infrastructure utilization (which GitHub is paying for) since it's building the same dependencies multiple times across jobs, whereas in a combined job they'd only be built once. Combined jobs would have a faster turnaround time when used with a single beefier self-hosted runner.

Otherwise, the only advantage I see to separating jobs by component is that you get a more granular view of the health of the project without having to drill into the job log.

Let me know if you'd like me to keep it this way or put it back into only being split out by GHC version.

**Comment by kquick** [comment]
Thanks for the investigation on this, @LisannaAtGalois .

I think the concept of being able to  adjust the module coverage on the PR's specific branch is a good way to handle that concern.

I do like the different modules as separate builds: I think it's easier to get a quick view of what parts are broken.  Otherwise if I'm working on semmc and it breaks semmc-ppc, but I'm not ready to fix semmc-ppc yet, I can easily see and ignore that aspect.  Otherwise I have to click details, expand the action that failed, then scroll down to find the error, which uses more of the user aspect (the highest cost element :-).

At this point, I think this looks good to me, and I assume you don't have anything else planned here.  I'd suggest getting a review from @travitch as well.

**Comment by travitch** [comment]
Thanks! I'm also very happy with this. Two questions/observations that don't need to block anything:
1. We might want to keep an eye on that `-j2` during the build; a few of the dependencies take a lot of memory to build and might blow up the low-memory runners
2. Do we have plans for a more centralized approach to building/distributing solver binaries? I've had to do similar things to get yices/z3 into dist tarballs before, and it feels really ad-hoc

**Comment by LisannaAtGalois** [comment]
> Thanks! I'm also very happy with this. Two questions/observations that don't need to block anything:
> 
>     1. We might want to keep an eye on that `-j2` during the build; a few of the dependencies take a lot of memory to build and might blow up the low-memory runners

OK - easy adjustment to make if we see it.

>     2. Do we have plans for a more centralized approach to building/distributing solver binaries? I've had to do similar things to get yices/z3 into dist tarballs before, and it feels really ad-hoc

@jared-w and I have been thinking of some ways to manage the logic for it, but I don't myself know about actually providing the binaries ourselves.

**Comment by travitch** [comment]
wooo

**Pull request merged**